jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketSessionState.java

```diff
@@ -123,6 +123,28 @@ public boolean onClosed(CloseStatus closeStatus)
         }
     }
 
+    /**
+     * This should only be called directly before closing the connection when we are in {@link State#CLOSED} state.
+     * This ensures an abnormal close status and if we have no error in the CloseStatus we will set one.
+     * @param t the error which occurred.
+     */
+    public void onError(Throwable t)
+    {
+        synchronized (this)
+        {
+            if (_sessionState != State.CLOSED || _closeStatus == null)
+                throw new IllegalArgumentException();
+
+            // Override any normal close status.
+            if (!_closeStatus.isAbnormal())
+                _closeStatus = new CloseStatus(CloseStatus.SERVER_ERROR, t);
+
+            // Otherwise set the error if it wasn't already set to notify onError as well as onClose.
+            if (_closeStatus.getCause() == null)
+                _closeStatus.initCause(t);
```

Comment:
this seams wrong.  If we were closed for some other reason, then adding a cause to it could put the wrong cause against a close reason.

Reply:
If we received an abnormal close frame and throw in `onFrame` then we want to keep the same close reason but we want to call `onError` as well so we need to add a cause to the close status so that `onError` is notified. I can actually change this line to get rid of the weird `initCause` method.
```java
_closeStatus = new CloseStatus(_closeStatus.getCode(), _closeStatus.getReason(), t);
```

Reply:
Definitely change to keep the close status immutable and get rid of the init.

But can you describe more.... maybe in comments... exactly why if we already have a close that we want to change just the cause and not the status and the reason as well?